File: KMSTraining.API/Services/AuthService.cs
using KMSTraining.API.Data;
using KMSTraining.API.DTOs;
using KMSTraining.API.Models;
using Microsoft.EntityFrameworkCore;

namespace KMSTraining.API.Services;

public class AuthService : IAuthService
{
    private readonly TripPlannerDbContext _context;
    private readonly ITokenService _tokenService;

    public AuthService(TripPlannerDbContext context, ITokenService tokenService)
    {
        _context = context;
        _tokenService = tokenService;
    }

    public async Task<AuthResponseDto> RegisterAsync(RegisterDto registerDto)
    {
        if (await _context.Users.AnyAsync(u => u.Username == registerDto.Username))
        {
            throw new InvalidOperationException("Username already exists");
        }

        if (await _context.Users.AnyAsync(u => u.Email == registerDto.Email))
        {
            throw new InvalidOperationException("Email already exists");
        }

        var passwordHash = BCrypt.Net.BCrypt.HashPassword(registerDto.Password);

        var user = new User
        {
            Username = registerDto.Username,
            Email = registerDto.Email,
            PasswordHash = passwordHash,
            FirstName = registerDto.FirstName,
            LastName = registerDto.LastName,
            CreatedAt = DateTime.UtcNow
        };

        _context.Users.Add(user);
        await _context.SaveChangesAsync();

        var token = _tokenService.GenerateToken(user.Id, user.Username, user.Email);

        return new AuthResponseDto
        {
            Token = token,
            Username = user.Username,
            Email = user.Email,
            ExpiresAt = DateTime.UtcNow.AddMinutes(60)
        };
    }

    public async Task<AuthResponseDto> LoginAsync(LoginDto loginDto)
    {
        var user = await _context.Users
            .FirstOrDefaultAsync(u => u.Username == loginDto.UsernameOrEmail || u.Email == loginDto.UsernameOrEmail);

        if (user == null || !BCrypt.Net.BCrypt.Verify(loginDto.Password, user.PasswordHash))
        {
            throw new UnauthorizedAccessException("Invalid credentials");
        }

        var token = _tokenService.GenerateToken(user.Id, user.Username, user.Email);

        return new AuthResponseDto
        {
            Token = token,
            Username = user.Username,
            Email = user.Email,
            ExpiresAt = DateTime.UtcNow.AddMinutes(60)
        };
    }
}

File: kmstraining.reactui/src/services/authService.ts
import api from './api';

export interface RegisterDto {
  username: string;
  email: string;
  password: string;
  firstName?: string;
  lastName?: string;
}

export interface LoginDto {
  usernameOrEmail: string;
  password: string;
}

export interface AuthResponse {
  token: string;
  username: string;
  email: string;
  expiresAt: string;
}

interface RawAuthResponse {
  token?: string;
  Token?: string;
  username?: string;
  Username?: string;
  email?: string;
  Email?: string;
  expiresAt?: string;
  ExpiresAt?: string;
}

export interface User {
  id: number;
  username: string;
  email: string;
  firstName?: string;
  lastName?: string;
  createdAt: string;
}

const normalizeAuthResponse = (data: RawAuthResponse): AuthResponse => {
  const token = data.token ?? data.Token;
  const username = data.username ?? data.Username;
  const email = data.email ?? data.Email;
  const expiresAt = data.expiresAt ?? data.ExpiresAt;

  if (!token || !username || !email || !expiresAt) {
    throw new Error('Invalid authentication response from server.');
  }

  return {
    token,
    username,
    email,
    expiresAt,
  };
};

const authService = {
  register: async (data: RegisterDto): Promise<AuthResponse> => {
    const response = await api.post<RawAuthResponse>('https://localhost:44391/api/Auth/register', data);
    return normalizeAuthResponse(response.data);
  },

  signIn: async (data: LoginDto): Promise<AuthResponse> => {
    const response = await api.post<RawAuthResponse>('https://localhost:44391/api/Auth/login', data);
    return normalizeAuthResponse(response.data);
  },

  login: async (data: LoginDto): Promise<AuthResponse> => {
    return authService.signIn(data);
  },

  logout: () => {
    localStorage.removeItem('token');
    localStorage.removeItem('user');
  },

  getCurrentUser: (): User | null => {
    const userStr = localStorage.getItem('user');
    return userStr ? JSON.parse(userStr) : null;
  },

  isAuthenticated: (): boolean => {
    return !!localStorage.getItem('token');
  },
};

export default authService;
